lib/std/channels.nim

```diff
@@ -0,0 +1,693 @@
+
+  close_fn = [
+    Mpmc: channelCloseMpmc,
+    Mpsc: channelCloseMpsc,
+    Spsc: channelCloseSpsc
+  ]
+
+  open_fn = [
+    Mpmc: channelOpenMpmc,
+    Mpsc: channelOpenMpsc,
+    Spsc: channelOpenSpsc
+  ]
+
+proc channelSend(chan: ChannelRaw, data: sink pointer, size: int, nonBlocking: bool): bool {.inline.} =
+  ## Send item to the channel (FIFO queue)
+  ## (Insert at last)
+  send_fn[chan.impl](chan, data, size, nonBlocking)
+
+proc channelReceive(chan: ChannelRaw, data: pointer, size: int, nonBlocking: bool): bool {.inline.} =
+  ## Receive an item from the channel
+  ## (Remove the first item)
+  recv_fn[chan.impl](chan, data, size, nonBlocking)
+
+proc channel_close(chan: ChannelRaw): bool {.inline.} =
+  ## Close a channel
+  close_fn[chan.impl](chan)
+
+proc channel_open(chan: ChannelRaw): bool {.inline.} =
+  ## (Re)open a channel
+  open_fn[chan.impl](chan)
+
+# Public API
+# ----------------------------------------------------------------------------------
+
+type
+  Chan*[T] = object ## Typed channels
+    d: ChannelRaw
+
+proc `=`[T](dest: var Chan[T]; src: Chan[T]) {.error.}
+
+proc `=destroy`[T](c: var Chan[T]) =
+  if c.d.buffer != nil: freeChannel(c.d)
+
+proc channelSend[T](chan: Chan[T], data: sink T, size: int, nonBlocking: bool): bool {.inline.} =
+  ## Send item to the channel (FIFO queue)
+  ## (Insert at last)
+  send_fn[chan.d.impl](chan.d, data.unsafeAddr, size, nonBlocking)
+
+proc channelReceive[T](chan: Chan[T], data: ptr T, size: int, nonBlocking: bool): bool {.inline.} =
+  ## Receive an item from the channel
+  ## (Remove the first item)
+  recv_fn[chan.d.impl](chan.d, data, size, nonBlocking)
+
+func trySend*[T](c: Chan[T], src: var Isolated[T]): bool {.inline.} =
+  ## Sends item to the channel(non blocking).
+  var data = src.extract
+  result = channelSend(c, data, sizeof(data), true)
+  if result:
+    wasMoved(data)
+
+template trySend*[T](c: Chan[T], src: T): bool =
+  trySend(c, isolate(src))
+
+func tryRecv*[T](c: Chan[T], dst: var T): bool {.inline.} =
+  ## Receives item from the channel(non blocking).
+  channelReceive(c, dst.addr, sizeof(dst), true)
+
+func send*[T](c: Chan[T], src: sink Isolated[T]) {.inline.} =
+  ## Sends item to the channel(blocking).
+  var data = src.extract
+  discard channelSend(c, data, sizeof(data), false)
+  wasMoved(data)
+
+template send*[T](c: var Chan[T]; src: T) =
+  send(c, isolate(src))
+
+func recv*[T](c: Chan[T], dst: var T) {.inline.} =
+  ## Receives item from the channel(blocking).
+  discard channelReceive(c, dst.addr, sizeof(dst), false)
+
+func recvIso*[T](c: Chan[T]): Isolated[T] {.inline.} =
+  var dst: T
+  discard channelReceive(c, dst.addr, sizeof(dst), false)
+  result = isolate(dst)
+
+func open*[T](c: Chan[T]): bool {.inline.} =
+  result = c.d.channel_open()
+
+func close*[T](c: Chan[T]): bool {.inline.} =
+  result = c.d.channel_close()
+
+func peek*[T](c: Chan[T]): int {.inline.} = peek(c.d)
+
+proc initChan*[T](elements = 30): Chan[T] =
```

Comment:
* IMO we shouldn't abbreviate it, `Channel[T]` is short enough (it's not like it's gonna appear thousands of times in your code)
* ditto, `initChan` => `initChannel`

Reply:
make sense

Reply:
fixed